retrievalmarket/impl/provider_environments.go

```diff
@@ -188,25 +221,87 @@ func (pde *providerDealEnvironment) CloseDataTransfer(ctx context.Context, chid
 func (pde *providerDealEnvironment) DeleteStore(storeID multistore.StoreID) error {
 	return pde.p.multiStore.Delete(storeID)
 }
-func getPieceInfoFromCid(pieceStore piecestore.PieceStore, payloadCID, pieceCID cid.Cid) (piecestore.PieceInfo, error) {
+
+func pieceInUnsealedSector(ctx context.Context, n retrievalmarket.RetrievalProviderNode, pieceInfo piecestore.PieceInfo) bool {
+	for _, di := range pieceInfo.Deals {
+		isUnsealed, err := n.IsUnsealed(ctx, di.SectorID, di.Offset.Unpadded(), di.Length.Unpadded())
+		if err != nil {
+			continue
+		}
+		if isUnsealed {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getAllDealsContainingPayload(pieceStore piecestore.PieceStore, payloadCID cid.Cid) ([]abi.DealID, error) {
 	cidInfo, err := pieceStore.GetCIDInfo(payloadCID)
 	if err != nil {
-		return piecestore.PieceInfoUndefined, xerrors.Errorf("get cid info: %w", err)
+		return nil, xerrors.Errorf("get cid info: %w", err)
+	}
+	var dealsIds []abi.DealID
+
+	for _, pieceBlockLocation := range cidInfo.PieceBlockLocations {
+		pieceInfo, err := pieceStore.GetPieceInfo(pieceBlockLocation.PieceCID)
+		if err != nil {
+			continue
+		}
+		for _, d := range pieceInfo.Deals {
+			dealsIds = append(dealsIds, d.DealID)
+		}
+	}
+
+	if len(dealsIds) == 0 {
+		return nil, xerrors.New("no deals found")
+	}
```

Comment:
If there was an error calling GetPieceInfo, and no other call to GetPieceInfo is successful, let's return the last error